### PR TITLE
Update broadcast_text_id in several tables

### PR DESCRIPTION
**Pull request by Lymphome**
Update the broadcast_text_id in 5 tables

## 1 Description
Cmangos teleported several text fields from original tables to broadcast_text table.
I built scripts to parse the old text fields and to compare them with the ones in the broadcast_text table.

Files inside this PR are the result of these scripts.
They produce sql queries and put some comments to verify the results.

## 2 Parsed tables
I let the full outputs in order to verify the contents.

1) 3 tables with same sql structure (for texts) :
* creature_ai_texts
        only 3 fields found (out of 28)
* dbscript_string
        18 fields found (out of 37)
* script_texts
        33 fields found (out of 156)

These tables just lack of some entries (about 250 for all the 3 tables), and the script for these tables don't find a lot of missing links but they are usefull to understand the generated output.

2) 2 tables with which need a lots of id :
* gossip_menu_option :
        Have 2 text fields (option and box) but the the second field concern only a very few entries. So the box text field is done by hand.
        3162 fields found (out of 3300)
        3224 fields are ok after "human read"

* npc_text_broadcast_text :
        Up to 8 text fields with their related probality
        This table takes much processing time as the script parses broadcast_text 8 times per npc_text entry.
        There is no sql delete query for the npc_text : the core is not happy of that but, anyhow, it runs.
        5656 entries completed (out of XX) 6471 text fields found (out of 6844)
        5891 entries and 6716 text fields are ok after "human read"

## 3 How to read the files

For each entry in source table, the script can display :

"# ## tab <_id> => <text>" : source id and text (what we have to search in broadcast table)
        for creature_ai_texts, dbscript_string and script_texts
"# ## tab <menu_id>/<_id> => <text>" : output for gossip_menu_option
        menu_id and id are the fields with the same name in table
"# ## txt <_position> (<proba>) => <text>" : output for npc_text
        for npc_text, the id of the row is displayed at beginning : "# ------- [<_id>]"
        foreach texts and proba in this row, it displays the 'position' of the field (text0, text1... text7)

"# == (mal|fem) [<_id>] "text"" : exact match found in broadcast table with its id and text (male or female text)
        generate an "UPDATE" query not commented
        the script stops its search on first match

"# ++ (mal|fem) [<_id>] "text"" : near exact match
        generate an "UPDATE" query, commented
        script searchs on whole broadcast table

"# -- (mal|fem) [<_id>] "text"" : search the same words
        only words are used to build a search regex
        no $n, $c...
        generate an "UPDATE" query, commented
        script searchs on whole broadcast table

"# oo (mal|fem) [<_id>] "text"" : search for certains words
        same as "# -- ", but one word can be omitted

"# NONE FOUND" : nothing revelant found
        does not mean that there is nothing (human look up necessary)
        creature_ai_texts, dbscript_string and script_texts do not have this

"# TOO MUCH" : too much matches (>20) found
        it occurs mainly in cases where we have one or two words in original text (like "train me", "Greeting"...)

"# FAR FOUND" : 'far' match found (only npc_text has this, only used for easy search for me)

"# HAND" : entry found and add by a human

Some entries have a commented "UPDATE" query with no id (ie "XXXXX") beacause I'm too lazy to copy/paste :)

Scripts are 'simple' pattern matchers, use brute force algo (nothing really smart), regex in a perl loop (perl is so nice for parsing texts)

## 4 misc
The files are valid sql files (mysql takes it fine)
Gossip_menu_option : end of file has some "HAND" entries for 'box' fields
Once the review is ok, I can clean the files for final PR (or just do a 'grep -E "^UPDATE"' or 'grep -A 2 -E "^INSERT"' for npc_text)
Npc_text table will need some "DELETE" because core complains about seeing texts in both npc_text and npc_text_broadcast_text
Some "exact"-match entries can be dangerous (like "I wish to unlearn my Talents", "Guild Master", "Weapons Trainer"... or other common texts)

### CODE REVIEW

**Comment by MantisLord** [comment]
Good work! will be reviewing this ASAP

**Comment by MantisLord** [comment]
Could you regenerate these files after you've run a query like this one? DELETE FROM broadcast_text WHERE Id > 27000;

Currently the TBC broadcast_text table contains way more entries than it should from later expansions. There may not be a clear defined cutoff point because it seems WotLK content broadcast texts start getting mixed in with late TBC content when you get into these high IDs. I chose 27000 here as a best guess when I stopped seeing TBC related things while scrolling through.

Again, very nice work. This is taking us closer to being able to drop all these tables ultimately...

**Comment by Lymphome** [comment]
In the scripts, there is a sql SELECT query on broadcast_text at start to get the data. I will add a WHERE Id < 27000 and rerun the scripts. I saw some entries with "wintergrasp", "panda..." and others while I was looking for failed searches too.

**Comment by offl** [comment]
> I chose 27000 here as a best guess when I stopped seeing TBC related things while scrolling through.

It will cut used texts, there are no low ids for some TBC or even classic texts
And 27000 is not enough for TBC

**Comment by AnonXS** [comment]
I also think they have the same design flaw as we have reusing lower ids for newer content, or reusing it.

**Comment by Lymphome** [comment]
I will change the ORDER in sql query for broadcast_text query (don't know why, it was DESC) => more low IDs in results.

I will add something to comment out high ID (> 27000 ?) and put a new tag ("# NOT TBC") in order to find them quickly.

**Comment by offl** [comment]
> I also think they have the same design flaw as we have reusing lower ids for newer content, or reusing it.

Mostly they add new id without touching old, that's why 99.9% of texts are still valid even if they were slightly changed(typos, tabs, grammar)
In ZA, Naxx, classic sometimes they changed current text, but those are pretty rare cases

File: Updates/0332_creature_ai_texts_broadcast_text_id.sql [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+
+# --------------------------------------------------------------------------------
+# ## tab -104	=> "Who challenges Naias? Puny $r, you are little better than those mindless trolls I have played against each other like so many pieces on a game board!"
+
+# == mal [13560] "Who challenges Naias? Puny $r, you are little better than those mindless trolls I have played against each other like so many pieces on a game board!"
+
+UPDATE creature_ai_texts SET broadcast_text_id = 13560 WHERE entry = -104;
+
+# --------------------------------------------------------------------------------
+# ## tab -321	=> "Iman m t'ief fu Fus'obeah italaf"
```
Comment (AnonXS) [inline Updates/0332_creature_ai_texts_broadcast_text_id.sql:13]: https://imgur.com/gallery/qHLnt

1910	Have a taste of Frostmane magics.	Have a taste of Frostmane magics.	14

**Comment by Lymphome** [comment]
About non-tbc text fields : 
* id 27711 is maybe the last id for tbc (apple vendor in shattrath, patch 2.4.3)
* some IDs seem to be after (brewfest, ahune, maybe others)

About double text fields : tests were not releavant

File: Updates/0332_creature_ai_texts_broadcast_text_id.sql (hunk 1 above)
Comment (Lymphome) [reply Updates/0332_creature_ai_texts_broadcast_text_id.sql:12]: Nice ! So, it is also possible to find some IDs with the "comment" field in tables : look for the npc quotes on wowhead, and sometimes in the comments from players.

Comment (offl) [reply Updates/0332_creature_ai_texts_broadcast_text_id.sql:12]: Or in creature_text table

**Comment by MantisLord** [comment]
This PR is almost fully merged aside from the npc_text_broadcast_text additions.
@Lymphome Can I bug you to re-run your script for this one table again? There have been alot of additions to this table from sniffs since this PR was first created.

**Comment by Lymphome** [comment]
Yes I will run asap. I have to merge several core updates before that (killerwife back with his eyes 8) )
Maybe in 2 or 3 weeks, I think...

File: Updates/0332_creature_ai_texts_broadcast_text_id.sql [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+# --------------------------------------------------------------------------------
+# ## tab -376	=> "[Orcish] Ogg kaz ko zugas kil mog Tov'nokaz Kagg"
+# COMMENT : "Orc Common Text"
+# NONE FOUND
+
+# --------------------------------------------------------------------------------
+# ## tab -377	=> "[Orcish] Tago mog regas gi mog Tov'nokaz Kagg"
+# COMMENT : "Orc Common Text"
+# NONE FOUND
+
```
Comment (RazzorX) [inline Updates/0332_creature_ai_texts_broadcast_text_id.sql:34]: Based on the translation into German
=> "[Orcish] Ogg kaz ko zugas kil mog Tov'nokaz Kagg" -376
=> "You are no match for the Blackrock Orcs!"  bid=1875

=> "[Orcish] Tago mog regas gi mog Tov'nokaz Kagg"  -377
=> "Feel the power of the Blackrock Orcs!"  bid=1874

but not 100% sure

File: Updates/0333_dbscript_string_broadcast_text_id.sql [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+# --------------------------------------------------------------------------------
+# ## tab 2000005032	=> "...And then the Tauren said, "13 INCHES!""
+# -- mal [6405] "And then the Tauren said, '13 INCHES!'"
+
+UPDATE dbscript_string SET broadcast_text_id = 6405 WHERE entry = 2000005032;
+
+# --------------------------------------------------------------------------------
+# ## tab 2000000215	=> "Apothecary Zamah smiles."
+# NONE FOUND
+
```
Comment (RazzorX) [inline Updates/0333_dbscript_string_broadcast_text_id.sql:14]: Could be
UPDATE dbscript_string SET broadcast_text_id = 1418 WHERE entry = 2000000215;

File: Updates/0332_creature_ai_texts_broadcast_text_id.sql (hunk 2 above)
Comment (AnonXS) [reply Updates/0332_creature_ai_texts_broadcast_text_id.sql:34]: https://github.com/cmangos/tbc-db/commit/5f6336d63c733cfcfdba572444dc172af46d3997#diff-8344129e35190a1fe0366238d7f8994431be263a6cfe54e26376f0479744540bR31291

File: Updates/0333_dbscript_string_broadcast_text_id.sql (hunk 3 above)
Comment (AnonXS) [reply Updates/0333_dbscript_string_broadcast_text_id.sql:14]: https://github.com/cmangos/tbc-db/blob/master/Updates/0859_dbscript_string_broadcast_text_update.sql#L55

dont bother too much alot is already done.

File: Updates/0333_dbscript_string_broadcast_text_id.sql [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+UPDATE dbscript_string SET broadcast_text_id = 12222 WHERE entry = 2000005545;
+
+# --------------------------------------------------------------------------------
+# ## tab 2000000166	=> "Sarath fades into the mist..."
+# FAR FOUND
+# oo mal [1476] "%s fades into the mist..."
+# oo mal [72076] "The ghost fades away into the mist..."
+
+# UPDATE dbscript_string SET broadcast_text_id = XXXXX WHERE entry = 2000000166;
+
```
Comment (RazzorX) [inline Updates/0333_dbscript_string_broadcast_text_id.sql:82]: I think bid=1476 is the right one. And as a side note, "Sarath" is wrong, he's called Sargath.

File: Updates/0335_gossip_menu_option_broadcast_text_id.sql [hunk 5]
```diff
@@ -0,0 +1,19919 @@
...
+UPDATE gossip_menu_option SET option_broadcast_text = 8261 WHERE id = 1 AND menu_id = 141;
+
+# --------------------------------------------------------------------------------
+# ## tab 141/4	=> "<Take the letter>"
+# NONE FOUND
+
+# --------------------------------------------------------------------------------
+# ## tab 161/1	=> "[PH] Get Thieves' Tools"
+# NONE FOUND
+
```
Comment (RazzorX) [inline Updates/0335_gossip_menu_option_broadcast_text_id.sql]: According to wotlk-db it should be bid=2621

UPDATE gossip_menu_option SET option_broadcast_text = 2621 WHERE id = 1 AND menu_id = 161;

**Comment by AnonXS** [comment]
Is any of this still applicable?

**Comment by AnonXS** [comment]
@MantisLord is anything still important here? i've seen we now need to do updates to BCT for ChatTypeID and stuff, maybe we need to extract the creature texts settings to paste them for BCT?

**Comment by MantisLord** [comment]
> @MantisLord is anything still important here? i've seen we now need to do updates to BCT for ChatTypeID and stuff, maybe we need to extract the creature ai texts settings to paste them for BCT?

Chat types from creature_ai_texts, script_texts, etc have already been applied to BCT table fields. This was done when the table was dropped.
Closing this PR now as everything major has been merged or overcome by events. Any additional BCT related fixes can go in a new PR.

**Pull request closed without merging**